File: paradigms/exercises/climbing_stairs/climbing_stairs.py
"""
Problem: Climbing Stairs
"""

def climb_stairs(n: int) -> int:
    if n <= 2:
        return n
    
    # Ways to reach step 1 and step 2
    two_steps_before = 1
    one_step_before = 2
    
    for _ in range(3, n + 1):
        current = one_step_before + two_steps_before
        two_steps_before = one_step_before
        one_step_before = current
        
    return one_step_before

if __name__ == "__main__":
    test_n = 5
    result = climb_stairs(test_n)
    print(f"Steps: {test_n}, Distinct ways: {result}")
    
    if climb_stairs(2) == 2 and climb_stairs(3) == 3 and result == 8:
        print("Test Passed! ✅")
    else:
        print("Test Failed! ❌")

File: paradigms/exercises/climbing_stairs/climbing_stairs.rs
/**
 * Problem: Climbing Stairs
 */

fn climb_stairs(n: i32) -> i32 {
    if n <= 2 {
        return n;
    }

    let mut two_steps_before = 1;
    let mut one_step_before = 2;

    for _ in 3..=n {
        let current = one_step_before + two_steps_before;
        two_steps_before = one_step_before;
        one_step_before = current;
    }

    one_step_before
}

fn main() {
    let test_n = 5;
    let result = climb_stairs(test_n);

    println!("Steps: {}, Distinct ways: {}", test_n, result);

    if climb_stairs(2) == 2 && climb_stairs(3) == 3 && result == 8 {
        println!("Test Passed! ✅");
    } else {
        println!("Test Failed! ❌");
    }
}
